### PR TITLE
F(by) talk and matching blog post: JuliaLang: The Ingredients for a Composable Programming Language

### PR DESCRIPTION
**Pull request by oxinabox**

### CODE REVIEW

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+But in the Julia Ecosystem, which was written mostly by Grad students, this seems to mostly work.
+This post explores some theories on why;
+and some lose recommendations to future language designers.
+<!--more-->
+
+
+This blog post is based on a talk I was invited to give at the 2020 [F(by) conference](fby.dev).
+Hopefully videos from that will be up soon, and I will link it here.
+This blog post is a bit ad-hoc in its ordering and content because of its origin as a talk, which was ad-hoc in ordering and content.
+I trust the reader will forgived me.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: typo

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+
+Parts of this post are inspired by [Stefan Karpinski's "The Unreasonable Effectiveness of Multiple Dispatch" talk at JuliaCon 2019](https://www.youtube.com/watch?v=kc9HwsxE1OY).
+I recommend that video, it goes into more details one some of the multiple dispatch points, and the subtle (but important) difference between julia's  dispatch and C++'s virtual methods.
+
+
+## What do I mean by composable ?
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: extra space before ?

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+
+### Examples:
+
+ - If you want to add tracking of measurement error to a scalar number, you shouldn't have to say anything about how your new type interacts with arrays **([Measurements.jl](https://github.com/JuliaPhysics/Measurements.jl))**
+ - If you have a Differential Equation solver, and a Neural Network library, then you should just be able to have neural ODEs (**[DifferentialEquations.jl](https://github.com/JuliaDiffEq/DifferentialEquations.jl) / [Flux.jl](https://github.com/FluxML/Flux.jl)**)
+ - If you have a package to add names to the dimensions of an array, and one to put arays on the GPU, then you shouldn't have to write code to have named arrays on the GPU (**[NamedDims.jl](https://github.com/invenia/NamedDims.jl) / [CUArrays.jl](https://github.com/JuliaGPU/CuArrays.jl)**)
+
+### Why is Julia this way?
+
+The theory I prose heremay sound counter-intuitive
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: here may

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+I suggest, that julia code is so reusable,
+are because the language has not just good features, but **weak and missing features.**
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Not sure how the sentence was intended, it sounds off to me.

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+only import what you need.
+e.g `using Foo: a, b c`
+
+Common practice in Julia is to do:
+`using Foo`,
+which imports everything everything that the author of `Foo` marked to be **exported**.
+
+You don't have to, but it's common.
+
+But what happens if one has package:
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *a package

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+using Bar
+training_data, test_data = ...
+mbar = BarModel(training_data)
+mfoo = FooModel(training_data)
+evaluate(predict(mbar), test_data)
+evaluate(predict(mfoo), test_data)
+```
+
+If you have multiple `using`s trying to bring the same name into scope,
+then julia throws an error.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *Julia

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+As a user you can tell it what to use.
+
+```julia
+evaluate(Bar.predict(mbar), test_data)
+evaluate(Foo.predict(mfoo), test_data)
+```
+
+### But the package authors can solve this:
+There is no name collision if both names *are* overloaded the from the same namespace.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Seems like a weird emphasis to me?

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+mbar = BarModel(training_data)
+mfoo = FooModel(training_data)
+evaluate(predict(mbar), test_data)
+evaluate(predict(mfoo), test_data)
+```
+
+
+### This encourages people to work together
+
+Name collisions makes package authors to come together and create base package (like `StatsBase`) and agree on what the functions mean.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Mix of singular and plural (collisions, makes),
package -> packages

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+
+They don't have to, since the user can still solve it, but it encourages it.
+Thus you get package authors thinking about other packages that might be used with theirs.
+
+One can even overload functions from multiple namespaces if you want;
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Maybe both times "you" or "one"?

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+
+The [recommended way](https://github.com/invenia/PkgTemplates.jl/) to create packages also ensures:
+
+ - Continuous Integration(/s) Setup
+ - Code coverage
+ - Documentation setup
+ - License set
+
+### Testing Julia code is important.
+Julia uses a JIT compiler, so  even compilation errors don't arive til run-time.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *arrive *until

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 11]
```diff
@@ -0,0 +1,627 @@
...
+As a dynamic language the type system says very little about correctness.
+
+Testing julia code is important.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *Julia

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+If code-paths are not covered in tests their is almost nothing in the language itself to protect them from having any kind of error.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *there

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 13]
```diff
@@ -0,0 +1,627 @@
...
+
+So its important to have Continuous Integration and other tooling all setup
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *it's
*set up
*.

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 14]
```diff
@@ -0,0 +1,627 @@
...
+
+### Making package making trival gets more packages made
+
+Recall that many Julia package authors are Grad students who are trying to get their next paper complete.
+Lots of scientific code never gets released, and lots of the code that does never gets made usable for others.
+But if it is easier to make a package, than it is to not, then it does.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Maybe there's another way to put this?

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 15]
```diff
@@ -0,0 +1,627 @@
...
+And once it is a package people start thinking more like package authors, and considering how it will be used.
+
+Its not a silver bullet but its one more push in the right direction.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *it's

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 16]
```diff
@@ -0,0 +1,627 @@
...
+## Multiple Dispatch + Duck-typing
+
+**Assume it walks like a duck and talks like a duck, and if it doesn't fix that.**
+
+---
+
+#### Aside: Open Classes
+Another closely related factor is **Open Classes.**
+But I'm not going to talk about that today, I recommend finding other resource to read on it.
+Such as [Eli Bendersky's blog post on the expression problem](https://eli.thegreenplace.net/2016/the-expression-problem-and-its-solutions#is-multiple-dispatch-necessary-to-cleanly-solve-the-expression-problem)
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Ending .

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 17]
```diff
@@ -0,0 +1,627 @@
...
+You need to allow new methods to be added to existing classes, in the first place.
+Some languages (e.g. Java) require that methods literally be places inside the same file as the class.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *placed

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 18]
```diff
@@ -0,0 +1,627 @@
...
+This means there is no way to add methods in another code-base, even unrelated ones.
+
+---
+
+### We would like to use some coode from a library
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *code

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 19]
```diff
@@ -0,0 +1,627 @@
...
+Consider I might have a type from the **Ducks** library.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+{% highlight julia %}
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: This doesn't render right.

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 20]
```diff
@@ -0,0 +1,627 @@
...
+    else
+        println("🐤 ➡️ 💧 Lead to water")
+    end
+end
+{% endhighlight %}
+</div>
+
+#### Rewriting the Duck has problems
+ - Have to edit someone elses library, to add support for *my* type.
+ - This could mean adding a lot of code for them to maintain
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Either end all points with "." or none?

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 21]
```diff
@@ -0,0 +1,627 @@
...
+ - Does not scale, what if other people wanted to add Chickens, Geese etc.
+
+##### Varient: Monkey-patch
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *Variant

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 22]
```diff
@@ -0,0 +1,627 @@
...
+ - If the language supports monkey patching, could do it that way
+ - but it means copying their code into my library, will run in to issues like not being able to update.
+ - Scaled to other people adding new types even worse, since no longer a central canonical source to copy
+
+##### Varient: could fork their code
+ - That is giving up on code reuse.
+
+### Option 2: Inherit from the Duck
+
+(NB: this example is **not** valid julia code)
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *Julia

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 23]
```diff
@@ -0,0 +1,627 @@
...
+end
+{% endhighlight %}
+</div>
+
+#### Inheriting from the Duck has problems:
+ - Have to replace every `Duck` in my code-base with `DuckWithSwanSupport`
+ - If I am using other libraries that might return a `Duck` I have to deal with that also
+
+##### Still does not scale:
+If someone else implements `DuckWithChickenSupport`, and I want to use both there code and mine, what do?
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: "what do I do"?

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 24]
```diff
@@ -0,0 +1,627 @@
...
+🦆 Quack
+🚶 Waddle
+🦢 Hiss
+🐤😢 Abandon
+🐤😢 Abandon
+{% endhighlight %}
+</div>
+
+
+## But does this happen in the wild?
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: Maybe specify you are not talking about ducks and swans in the wild ;-)

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 25]
```diff
@@ -0,0 +1,627 @@
...
+
+Turns out it does.
+
+The need to extend operations to act on new combinations of types shows up all the time in scientific computing.
+I suspect it shows up more generally, but we've learned to ignore it.
+
+
+If you look at a list of BLAS, methods you will see just this, encoded in the function name
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: remove ,

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 26]
```diff
@@ -0,0 +1,627 @@
...
+ - Detect important cases via tracing
+ - Compile specialized methods for them
+
+This is called specialization.
+
+### Basic functionalitionality of Julia's JIT:
+
+ - Specialize all methods on all types that they are called on as they are called
+
+This is pretty good: its a reasonable assumption that the types are going to an important case.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *it's

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 27]
```diff
@@ -0,0 +1,627 @@
...
+ - Symbolic Algebra numbers
+
+Without ever having to touch array code, except as a late-stage optimization.
+
+Otherwise, one needs to implement array support into one's scalars, to have reasonable performance at all.
+
+## It is good to invent new languages
+
+People need to invent new languages.
+Its a good time to be inventing new languages.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: It's

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 28]
```diff
@@ -0,0 +1,627 @@
...
+It's good for the world.
+Its good for me because I like cool new things.
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: It's

File: _posts/2020-02-08-whycompositionaljulia.md [hunk 29]
```diff
@@ -0,0 +1,627 @@
...
+
+I’ld just really like those new languages to please have:
```
Comment (karajan9) [inline _posts/2020-02-08-whycompositionaljulia.md]: *I'd

**Review by karajan9 — CHANGES_REQUESTED** [review]
I really like the content and think it is well thought out. The blog still reads to me more like talk slides than a blog post, maybe it would be beneficial to form it into a more flowing text?

File: _posts/2020-02-08-whycompositionaljulia.md (hunk 7 above)
Comment (oxinabox) [reply _posts/2020-02-08-whycompositionaljulia.md]: Yes, looks like a left over from a earlier version of this sentenced.
Rephrases and moved empisis

File: _posts/2020-02-08-whycompositionaljulia.md (hunk 11 above)
Comment (oxinabox) [reply _posts/2020-02-08-whycompositionaljulia.md]: I am going to stick to lower-case julia
It makes it less like a person's name.

Comment (karajan9) [reply _posts/2020-02-08-whycompositionaljulia.md]: I remember reading somewhere that the name of the language is supposed to be "Julia" and "julia" refers to the executable. I can't find it right now though, so I guess it's fine either way ;-)

File: _posts/2020-02-08-whycompositionaljulia.md (hunk 19 above)
Comment (oxinabox) [reply _posts/2020-02-08-whycompositionaljulia.md]: It does in published form. Those are some templating language that Jekyll uses.
(It didn't use to support syntax highlighting in any other way, but now I think it does. However my scripts still generate the old format)

It just doesn't render in GitHub.

**Comment by oxinabox** [comment]
> The blog still reads to me more like talk slides than a blog post, maybe it would be beneficial to form it into a more flowing text?

Having done that before it takes more time than I have right now.

The transition from: [these slides](https://rawcdn.githack.com/oxinabox/oxinabox.github.io/2cab1a7528cd5e2395de03ffc7767a4006d5ea15/_drafts/PyDataCambridgeMeetup2019/JuliaLangFeatures.slides.html) to the pair of Emergent Features of Julia Lang blog posts:
https://invenia.github.io/blog/2019/10/30/julialang-features-part-1/
and https://invenia.github.io/blog/2019/11/06/julialang-features-part-2/
took multiple people-days.

Which is why this is going out on my blog rather than Invenia's.
Because the company has a different kind of reputation to maintain.

**Comment by karajan9** [comment]
Fair enough! Thanks for doing this in the first place.

**Comment by cormullion** [comment]
Is there a latest version that I can look through?

**Comment by oxinabox** [comment]
> Is there a latest version that I can look through?

Latest version  is what I pushed.
I don't have an easy way to show previews as it will look in final.
(What I do is spin up local jekyll)

**Comment by cormullion** [comment]
I'll give it a read now... give me till midday :)

**Comment by cormullion** [comment]
Here's a few corrections:

11 someone else's

26 details on some of

92 overloaded are the from ??

163 enclined => inclined

183 benifits => benefits

194 finding other resource to read on it => other resources to read up about it?

450 someone elses library => someone else's

582  its a => it's

641 what ever  => whatever 

We need a better way to do proof-reading! :)

**Comment by oxinabox** [comment]
Thanks all for the help, releasing now

**Pull request merged**